### PR TITLE
abi: import files from latest geth release

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes
**This PR updates abi package and abigen. It applies all abi related changes up to v1.9.21.**

abi, bind package is identical with geth v1.9.21. However, backends package is not exactly identical because some geth changes are not applied yet. I didn’t applied them or applied part of them because those changes are huge and not only included at abi package. Also, I removed unnecessary codes such as vyper, clef, etc. You can refer the **commit msgs** which geth PR is applied. 

**Below PRs are the PR lists that i didn’t included. If needed, please refer them.**
* internal/ethapi: return revert reason (geth [#21083](https://github.com/ethereum/go-ethereum/pull/21083))
* all: seperate consensus error and evm internal error (geth [#20830](https://github.com/ethereum/go-ethereum/pull/20830))
* eth/filters, interfaces.go: EIP-234 Add blockHash option to eth_getLogs (geth [#16734](https://github.com/ethereum/go-ethereum/pull/16734))
* accounts/abi/bind/backend, internal/ethapi: recap gas limit with balance  (geth [#21043](https://github.com/ethereum/go-ethereum/pull/21043), geth [#21346](https://github.com/ethereum/go-ethereum/pull/21346))

**fix NewFallback TC**
NewFallback TC contains a bug. Fixed TC and fixed NewFallback contract to below. 
Related Geth PR https://github.com/ethereum/go-ethereum/pull/22053
```
pragma solidity ^0.6.2;

contract NewFallbacks {
  event Fallback(bytes data);
  fallback() external { 
      emit Fallback(msg.data);
  }
  event Received(address addr, uint value); 
  receive() external payable {
      emit Received(msg.sender, msg.value);
  }
}
```

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by ehnuje — COMMENTED** [review]
It would be appreciated if you describe the which release you used to make this PR. Also, did you check the code diff result of this code and the code in the geth? Are they identical?

**Comment by yoomee1313** [comment]
> It would be appreciated if you describe the which release you used to make this PR. Also, did you check the code diff result of this code and the code in the geth? Are they identical?

This is good idea. I'll make a description about the release information. I checked the diff result, but it's not identical because some PRs are not applied since they are related with other packages. However, at that time, i thought it wasn't necessary to bring those PRs, but now, it would be better to bring them to match the codebase as much as possible. I'll let you know after more work.

File: common/compiler/solidity.go [hunk 1]
```diff
@@ -187,15 +180,3 @@ func (s *Solidity) run(cmd *exec.Cmd, source string) (map[string]*Contract, erro
 	}
 	return contracts, nil
 }
-
```
Comment (KimKyungup) [inline common/compiler/solidity.go:190]: This is moved to common/compiler/helpers.go. :)

**Review by KimKyungup — COMMENTED** [review]
I compared this abi pkg code with geth reference code with @yoomee1313.
There is some difference. But @yoomee1313 will list up the PRs which is not applied in this PR. They caused some differences.

And I reviewed extra code also. I think it's ok.

However, the new fallback test code doesn't work.
@yoomee1313 Please check the test. 
I think there may be some differences in the transaction call routine.

File: accounts/abi/bind/bind_test.go [hunk 2]
```diff
@@ -950,12 +1066,635 @@ var bindTests = []struct {
...
+			sim.Commit()
+	
+			var gotEvent bool
+			iter, _ := c.FilterReceived(nil)
+			defer iter.Close()
+			for iter.Next() {
+				if iter.Event.Addr != addr {
+					t.Fatal("Msg.sender mismatch")
+				}
+				if iter.Event.Value.Uint64() != 100 {
```
Comment (KimKyungup) [inline accounts/abi/bind/bind_test.go:1658]: How about theses?
```suggestion
				if iter.Event.Value.Uint64() != opts.Value.Uint64() {
```

```suggestion
				if iter.Event.Value.Cmp(opts.Value) != 0 {
```

File: common/compiler/solidity.go (hunk 1 above)
Comment (yoomee1313) [reply common/compiler/solidity.go:190]: @KimKyungup yes, geth have made helpers.go file and moved slurpFiles to the helpers.go file.

File: accounts/abi/bind/bind_test.go (hunk 2 above)
Comment (yoomee1313) [reply accounts/abi/bind/bind_test.go]: Thank you, i applied it.

**Review by KimKyungup — APPROVED** [review]

**Comment by KimKyungup** [comment]
In content, all geth PR link is broken. Could you link them with their PRs?

**Review by ehnuje — DISMISSED** [review]

**Comment by ehnuje** [comment]
@yoomee1313 Could you please share the progress of this PR? Are there any other things left to merge this change?

**Comment by yoomee1313** [comment]
Forgive me for being late. I was delayed trying to find out about the situations below.
(1) NewFallback fix verification has been completed.
-> In the original code, calldata was copied to memory using calldatacopy and then copied to the event through parameters.
-> However, there was a problem in the above process, so the structure has been changed to copy `msg.data` to the event through parameters.
(2) The execution of the OpenZeppelin Test Suite has been completed.
-> In the openzeppelin test, it was checked whether this PR had an effect on the revert reason error. The error is that the revert reason is not returned.
-> The conclusion is that the error is not affected by this PR. Instead, it is related to the PRs (especially eth_call revert reason) which are described in the above description. We will complete the support for this in the future with next `incompatibility change`.

In conclusion, this PR supports `solidity 0.6` including New Fallback and struct type.
Please take a look! @KlaytnDev

**Review by KimKyungup — APPROVED** [review]

**Review by ehnuje — APPROVED** [review]

**Comment by KimKyungup** [comment]
PTAL @aidan-kwon. Let's merge~!!!

**Review by aidan-kwon — APPROVED** [review]
